File: javascript/binary-search/binary-search.js
export const find = (haystack, needle) => {
  let lo = 0, hi = haystack.length - 1;
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (haystack[mid] > needle) {
      hi = mid - 1;
    } else if (haystack[mid] < needle) {
      lo = mid + 1
    } else return mid;
  }
  throw new Error('Value not in array');
};

File: lua/binary-search/binary-search.lua
return function(array, target)
  local lo = 1
  local hi = #array + 1
  while lo < hi do
    local mid = math.floor((lo + hi) / 2)
    if array[mid] > target then
      hi = mid
    elseif array[mid] < target then
      lo = mid + 1
    else
      return mid
    end
  end
  return -1
end
